### PR TITLE
Add GDAL python repository workflow

### PR DESCRIPTION
**Pull request by alexberndt**
### Changes

Adding a shared workflow for usage in our GDAL python repositories.

### Motivation

Adding improvements e.g. caching, clarifying docs, new features take a lot of time because each GDAL python repo has their own, hardcoded CI workflow.

### Notable Features

- **Check SDK and package version match**
   Flag which tests if the SDK version and package version match (`poetry version -s`)
- **2 Test "Stages"**
    
    Ability to run two individual sets of test combinations. 

    Examples of where these two stages are useful:
    - Run stage 1 (fast) tests in CI on each commit, stage 2 (slow) tests only on merge to `main` (currently needed in`data-ingestion`, `coregistration` and `image-inventory`)
    - Run stage 1 (parallel) tests in parallel using `pytest-xdist` and stage 2 (dask) tests sequentially to avoid conflicts between `dask` and `pytest-xdist` (currently needed in `risk-score`)

### Scope of this PR

1. Create workflow to be used in multiple GDAL/Python repositories

2. Test this workflow on the following repositories:

    - [x] risk-score: https://github.com/20treeAI/risk_score/pull/411
    - [x] data-ingestion: https://github.com/20treeAI/data-ingestion/pull/539
    - [x] coregistration: https://github.com/20treeAI/coregistration/pull/155
    - [x] image-inventory: https://github.com/20treeAI/image-inventory/pull/86

### CODE REVIEW

**Comment by alexberndt** [comment]
[Add shared workflow for Python GDAL repositories](https://trello.com/c/rvFIpW6p/396-add-shared-workflow-for-python-gdal-repositories)

File: .github/workflows/python_gdal_ci.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+      # TESTS Stage 1 and Stage 2
+      - name: Import GCS key for tests
+        run: echo "${{ secrets.TEST_SERVICEACCOUNT_KEY }}" | base64 -d > $GOOGLE_APPLICATION_CREDENTIALS
+
+      - if: inputs.test_stage_1_should_run
+        name: Run Stage 1 Tests
+        run: | 
+          if [ ${{ inputs.test_stage_1_use_pytest_xdist }} ]; then
+            pytest --durations=0 -m '${{ inputs.test_stage_1_conditions }}' -n auto --disable-warnings
```
Comment (roelofvandijk) [inline .github/workflows/python_gdal_ci.yml:163]: I'm not sure we should use xdist for everything - it makes runs faster but can potentially hide issues with multithreaded geozarr loading...

File: .github/workflows/python_gdal_ci.yml [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        required: false
+      TEST_SERVICEACCOUNT_KEY:
+        description: 'Service account key with access to resources needed for running tests'
+        required: true
+
+
+
+jobs:
+  ci:
+    runs-on: ubuntu-latest
```
Comment (roelofvandijk) [inline .github/workflows/python_gdal_ci.yml:85]: Should we run this on a gdal docker image? This would not test the env creation anymore...

**Comment by roelofvandijk** [comment]
Why not use/expand the existing python workflow?

**Comment by roelofvandijk** [comment]
One issue: assumes xdist installed, will break if it isn't

**Comment by roelofvandijkO** [comment]
We could use this to enforce one GDAL version everywhere, by adding a job that parses the `env.yml` file and checks if the GDAL version is the expected GDAL version.

File: .github/workflows/python_gdal_ci.yml [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        required: true
+        type: string
+      python_version:
+        description: 'Python version for building lib'
+        default: '3.10.6'
+        required: false
+        type: string
+      poetry_version:
+        description: 'Poetry version for building lib'
+        default: '1.5.1'
```
Comment (MaxLenormand) [inline .github/workflows/python_gdal_ci.yml:20]: Any reason not to use latest, i.e. `1.7`?

**Review by MaxLenormand — APPROVED** [review]
I don't have enough context for all the implementation details, but from what I see LGTM

I'll leave it up to others to also review

File: .github/workflows/python_gdal_ci.yml (hunk 1 above)
Comment (alexberndt) [reply .github/workflows/python_gdal_ci.yml:163]: Indeed, this is why I added the `test_stage_1_use_pytest_xdist` and `test_stage_2_use_pytest_xdist` variables so teams can either use or not use `pytest-xdist` depending on their use case, as documented [here](https://github.com/20treeAI/github-workflows/pull/79/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R267-R276)

File: .github/workflows/python_gdal_ci.yml (hunk 2 above)
Comment (alexberndt) [reply .github/workflows/python_gdal_ci.yml:85]: For now, I'm just trying to replicate our current CIs in Python GDAL repos, which all use ubuntu-latest. I would consider running within GDAL docker images as a next step.

**Comment by alexberndt** [comment]
> Why not use/expand the existing python workflow?

Do you mean the [(SDK) library workflow](https://github.com/20treeAI/github-workflows/blob/main/.github/workflows/python_library_ci.yml)? I think I'd keep them separate, at least for now, since most repos use both, e.g. risk-score's [sdk library CI](https://github.com/20treeAI/risk_score/blob/main/.github/workflows/sdk_ci_pipeline.yml) and [regular CI](https://github.com/20treeAI/risk_score/blob/main/.github/workflows/ci_pipeline.yml)

We could consider merging them, but let's consider this as a next step. This CI already has many variables, so it might get convoluted/confusing.

> One issue: assumes xdist installed, will break if it isn't

By setting `test_stage_1_use_pytest_xdist=false` and `test_stage_2_use_pytest_xdist=false`, the CI uses regular `pytest` for both test stages, so there shouldn't be a problem if `xdist` is not installed in a particular repo.

> We could use this to enforce one GDAL version everywhere, by adding a job that parses the env.yml file and checks if the GDAL version is the expected GDAL version.

Maybe for a next step. Enforcing GDAL versions across the entire codebase is tricky because sometimes people take longer to upgrade GDAL (due to breaking changes).

**Comment by roelofvandijkO** [comment]
@alexberndt one approach for xdist would be to use it if present, instead of using the flag.

Something llike a (bash) ternary in the pytest command`pytest {{ $(poetry run pip list | grep xdist) ? "-n auto" : "" }}`

**Pull request merged**